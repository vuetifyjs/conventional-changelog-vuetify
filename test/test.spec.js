import { describe, it, expect } from 'vitest'
const conventionalChangelogCore = require('conventional-changelog-core')
const preset = require('../')
const gitDummyCommit = require('git-dummy-commit')
const shell = require('shelljs')
const betterThanBefore = require('better-than-before')()
const preparing = betterThanBefore.preparing

function streamToString (stream) {
  const chunks = []
  return new Promise((resolve, reject) => {
    stream.on('data', (chunk) => chunks.push(Buffer.from(chunk)))
    stream.on('error', (err) => reject(err))
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')))
  })
}

betterThanBefore.setups([
  () => {
    shell.config.resetForTesting()
    shell.cd(__dirname)
    shell.rm('-rf', 'tmp')
    shell.mkdir('tmp')
    shell.cd('tmp')
    shell.mkdir('git-templates')
    shell.exec('git init --template=./git-templates')

    gitDummyCommit(['build: first build setup', 'BREAKING CHANGE: New build system.'])
    gitDummyCommit(['ci(travis): add TravisCI pipeline', 'BREAKING CHANGE: Continuously integrated.'])
    gitDummyCommit(['feat: amazing new module', 'BREAKING CHANGE: Not backward compatible.'])
    gitDummyCommit(['fix(compile): avoid a bug', 'BREAKING CHANGE: The Change is huge.'])
    gitDummyCommit(['perf(ngOptions): make it faster', ' closes #1, #2'])
    gitDummyCommit('revert(ngOptions): bad commit')
    gitDummyCommit('fix: oops')
  },
  () => {
    gitDummyCommit(['feat(awesome): addresses the issue brought up in #133'])
  },
  () => {
    gitDummyCommit(['feat(awesome): fix #88'])
  },
  () => {
    gitDummyCommit(['feat(awesome): issue brought up by @bcoe! on Friday'])
  },
  () => {
    gitDummyCommit(['build(npm): edit build script', 'BREAKING CHANGE: The Change is huge.'])
    gitDummyCommit(['ci(travis): setup travis', 'BREAKING CHANGE: The Change is huge.'])
    gitDummyCommit(['docs(readme): make it clear', 'BREAKING CHANGE: The Change is huge.'])
    gitDummyCommit(['style(whitespace): make it easier to read', 'BREAKING CHANGE: The Change is huge.'])
    gitDummyCommit(['refactor(code): change a lot of code', 'BREAKING CHANGE: The Change is huge.'])
    gitDummyCommit(['test(*): more tests', 'BREAKING CHANGE: The Change is huge.'])
  },
  () => {
    shell.exec('git tag v1.0.0')
    gitDummyCommit('feat: some more features')
  },
  () => {
    gitDummyCommit(['feat(*): implementing #5 by @dlmr', ' closes #10'])
  },
  () => {
    gitDummyCommit(['Revert \\"feat: default revert format\\"', 'This reverts commit 1234.'])
    gitDummyCommit(['revert: feat: custom revert format', 'This reverts commit 5678.'])
  },
])

expect.addSnapshotSerializer({
  serialize (val) {
    let i = 0
    const commits = val.matchAll(/\/commit\/([a-z0-9]{40})/g)
    return Array.from(commits).reduce((acc, v) => {
      ++i
      return acc.replaceAll(
        `[${v[1].substring(0, 7)}](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/${v[1]})`,
        `[${String(i).padStart(7, '0')}](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/${String(i).padStart(40, '0')})`,
      )
    }, val)
  },

  test (val) {
    return typeof val === 'string'
  },
})

describe('vuetify preset', function () {
  it('should work if there is no semver tag', async () => {
    preparing(1)

    const changelog = await streamToString(
      conventionalChangelogCore({ config: preset })
    )

    expect(changelog).toMatchSnapshot()
  })

  it('should replace #[0-9]+ with GitHub issue URL', async () => {
    preparing(2)

    const changelog = await streamToString(
      conventionalChangelogCore({ config: preset })
    )

    expect(changelog).toMatchSnapshot()
  })

  it('should remove the issues that already appear in the subject', async () => {
    preparing(3)

    const changelog = await streamToString(
      conventionalChangelogCore({ config: preset })
    )

    expect(changelog).toMatchSnapshot()
  })

  it('should replace @username with GitHub user URL', async () => {
    preparing(4)

    const changelog = await streamToString(
      conventionalChangelogCore({ config: preset })
    )

    expect(changelog).toMatchSnapshot()
  })

  it('should not discard commit if there is BREAKING CHANGE', async () => {
    preparing(5)

    const changelog = await streamToString(
      conventionalChangelogCore({ config: preset })
    )

    expect(changelog).toMatchSnapshot()
  })

  it('should work if there is a semver tag', async () => {
    preparing(6)

    const changelog = await streamToString(
      conventionalChangelogCore({
        config: preset,
        outputUnreleased: true,
      })
    )

    expect(changelog).to.include('some more features')
    expect(changelog).to.not.include('BREAKING')
  })

  it('should only replace with link to user if it is an username', async () => {
    preparing(7)

    const changelog = await streamToString(
      conventionalChangelogCore({ config: preset })
    )

    expect(changelog).toMatchSnapshot()
  })

  it('parses both default (Revert "<subject>") and custom (revert: <subject>) revert commits', async () => {
    preparing(8)

    const changelog = await streamToString(
      conventionalChangelogCore({ config: preset })
    )

    expect(changelog).toMatchSnapshot()
  })
})
