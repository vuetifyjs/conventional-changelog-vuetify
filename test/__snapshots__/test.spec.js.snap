// Vitest Snapshot v1

exports[`vuetify preset > parses both default (Revert "<subject>") and custom (revert: <subject>) revert commits 1`] = ``;

exports[`vuetify preset > should not discard commit if there is BREAKING CHANGE 1`] = `
# 1.0.0 (2022-02-22)


### Tests

* more tests ([0000001](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000001))


### Code Refactoring

* **code:** change a lot of code ([0000002](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000002))


### Styles

* **whitespace:** make it easier to read ([0000003](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000003))


### Documentation

* **readme:** make it clear ([0000004](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000004))


### Continuous Integration

* **travis:** add TravisCI pipeline ([0000005](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000005))
* **travis:** setup travis ([0000006](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000006))


### Build System

* first build setup ([0000007](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000007))
* **npm:** edit build script ([0000008](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000008))


### :rocket: Features

* amazing new module ([0000009](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000009))
* **awesome:** addresses the issue brought up in [#133](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/133) ([0000010](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000010))
* **awesome:** fix [#88](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/88) ([0000011](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000011))
* **awesome:** issue brought up by [@bcoe](https://github.com/bcoe)! on Friday ([0000012](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000012))


### :wrench: Bug Fixes

* **compile:** avoid a bug ([0000013](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000013))
* oops ([0000014](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000014))


### :arrows_counterclockwise: Reverts

* **ngOptions:** bad commit ([0000015](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000015))


### Performance Improvements

* **ngOptions:** make it faster ([0000016](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000016)), closes [#1](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/1) [#2](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/2)


### BREAKING CHANGES

* The Change is huge.
* **code:** The Change is huge.
* **whitespace:** The Change is huge.
* **readme:** The Change is huge.
* **travis:** The Change is huge.
* **npm:** The Change is huge.
* **compile:** The Change is huge.
* Not backward compatible.
* **travis:** Continuously integrated.
* New build system.




`;

exports[`vuetify preset > should only replace with link to user if it is an username 1`] = ``;

exports[`vuetify preset > should remove the issues that already appear in the subject 1`] = `
# 1.0.0 (2022-02-22)


### :rocket: Features

* amazing new module ([0000001](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000001))
* **awesome:** addresses the issue brought up in [#133](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/133) ([0000002](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000002))
* **awesome:** fix [#88](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/88) ([0000003](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000003))


### :wrench: Bug Fixes

* **compile:** avoid a bug ([0000004](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000004))
* oops ([0000005](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000005))


### :arrows_counterclockwise: Reverts

* **ngOptions:** bad commit ([0000006](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000006))


### Performance Improvements

* **ngOptions:** make it faster ([0000007](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000007)), closes [#1](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/1) [#2](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/2)


### Continuous Integration

* **travis:** add TravisCI pipeline ([0000008](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000008))


### Build System

* first build setup ([0000009](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000009))


### BREAKING CHANGES

* **compile:** The Change is huge.
* Not backward compatible.
* **travis:** Continuously integrated.
* New build system.




`;

exports[`vuetify preset > should replace #[0-9]+ with GitHub issue URL 1`] = `
# 1.0.0 (2022-02-22)


### :rocket: Features

* amazing new module ([0000001](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000001))
* **awesome:** addresses the issue brought up in [#133](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/133) ([0000002](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000002))


### :wrench: Bug Fixes

* **compile:** avoid a bug ([0000003](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000003))
* oops ([0000004](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000004))


### :arrows_counterclockwise: Reverts

* **ngOptions:** bad commit ([0000005](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000005))


### Performance Improvements

* **ngOptions:** make it faster ([0000006](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000006)), closes [#1](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/1) [#2](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/2)


### Continuous Integration

* **travis:** add TravisCI pipeline ([0000007](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000007))


### Build System

* first build setup ([0000008](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000008))


### BREAKING CHANGES

* **compile:** The Change is huge.
* Not backward compatible.
* **travis:** Continuously integrated.
* New build system.




`;

exports[`vuetify preset > should replace @username with GitHub user URL 1`] = `
# 1.0.0 (2022-02-22)


### :rocket: Features

* amazing new module ([0000001](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000001))
* **awesome:** addresses the issue brought up in [#133](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/133) ([0000002](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000002))
* **awesome:** fix [#88](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/88) ([0000003](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000003))
* **awesome:** issue brought up by [@bcoe](https://github.com/bcoe)! on Friday ([0000004](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000004))


### :wrench: Bug Fixes

* **compile:** avoid a bug ([0000005](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000005))
* oops ([0000006](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000006))


### :arrows_counterclockwise: Reverts

* **ngOptions:** bad commit ([0000007](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000007))


### Performance Improvements

* **ngOptions:** make it faster ([0000008](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000008)), closes [#1](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/1) [#2](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/2)


### Continuous Integration

* **travis:** add TravisCI pipeline ([0000009](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000009))


### Build System

* first build setup ([0000010](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000010))


### BREAKING CHANGES

* **compile:** The Change is huge.
* Not backward compatible.
* **travis:** Continuously integrated.
* New build system.




`;

exports[`vuetify preset > should work if there is no semver tag 1`] = `
# 1.0.0 (2022-02-22)


### :wrench: Bug Fixes

* **compile:** avoid a bug ([0000001](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000001))
* oops ([0000002](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000002))


### :arrows_counterclockwise: Reverts

* **ngOptions:** bad commit ([0000003](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000003))


### Performance Improvements

* **ngOptions:** make it faster ([0000004](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000004)), closes [#1](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/1) [#2](https://github.com/vuetifyjs/conventional-changelog-vuetify/issues/2)


### :rocket: Features

* amazing new module ([0000005](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000005))


### Continuous Integration

* **travis:** add TravisCI pipeline ([0000006](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000006))


### Build System

* first build setup ([0000007](https://github.com/vuetifyjs/conventional-changelog-vuetify/commit/0000000000000000000000000000000000000007))


### BREAKING CHANGES

* **compile:** The Change is huge.
* Not backward compatible.
* **travis:** Continuously integrated.
* New build system.




`;
